api/envoy/extensions/filters/http/compressor/v3/compressor.proto

```diff
@@ -96,7 +96,8 @@ message Compressor {
       [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
 
   // Runtime flag that controls whether the filter is enabled or not. If set to false, the
-  // filter will operate as a pass-through filter. If not specified, defaults to enabled.
+  // filter will operate as a pass-through filter, unless overridden by
+  // CompresserPerRoute.response_compression_enabled. If not specified, defaults to enabled.
```

Comment:
I agree that overriding the semantics may not be the best approach. That said, I don't see this as a breaking change, because one would need to use the new `CompressorPerRoute` config, which is only introduced now.

Rejecting the config is one possibility, and if chosen needs to be clearly clarified in the `CompressorPerRoute`.

Maybe to keep it consistent with other PerRoute configs, why not have the following field:
`bool disabled = 1 [(validate.rules).bool = {const: true}];`
That only allows to disable, if the filter is configured to allow (with runtime flag enabled)?

Reply:
I need a way to have response compression only enabled on routes that want it. We have a multi-tenant deployment with tons of routes. We'd much rather push new config to the few routes that opt in to compression than the great many routes that don't want it.

With this proposal, we can set Compressor.response_direction_config.enabled.default_value to false, then add a CompresserPerRoute config to only the routes that want compression.

Reply:
> I agree that overriding the semantics may not be the best approach. That said, I don't see this as a breaking change, because one would need to use the new `CompressorPerRoute` config, which is only introduced now.
> 

I'm afraid to break the future, since `runtime_enabled` will be removed in the future (assume someday we may bump the major version), then the new field shouldn't have different behavior based on deprecated behavior. But yes, change again, only when the field not specified the behavior will changed, it should be fine. But yes, still agree overriding semantics doesn't seem good here.

> Rejecting the config is one possibility, and if chosen needs to be clearly clarified in the `CompressorPerRoute`.
> 

`runtime_enabled ` seems like a flag controlling whether the whole filter is enabled. 

> Maybe to keep it consistent with other PerRoute configs, why not have the following field: `bool disabled = 1 [(validate.rules).bool = {const: true}];` That only allows to disable, if the filter is configured to allow (with runtime flag enabled)?

+1 `disabled`

Reply:
Looking again at this API, it seems that the `runtime_enabled` was deprecated in favor of a [per-direction runtime flag](https://github.com/envoyproxy/envoy/blob/e4ae2259add5a327c4d67de18a82df5d78a801b4/api/envoy/extensions/filters/http/compressor/v3/compressor.proto#L34).
I think it complicates things a bit, if we just want to add a BoolValue to enable/disable compression.

As an alternative, will it make sense to override the RuntimeFeatureFlag? this way one can have multiple paths enabled/disabled by setting a single flag.

Reply:
> I'm afraid to break the future, since runtime_enabled will be removed in the future (assume someday we may bump the major version), then the new field shouldn't have different behavior based on deprecated behavior.

When `runtime_enabled` is removed, the effect of `response_compression_enabled` will not change in any way. If a `CompressorPerRoute` is present, `runtime_enabled` is ignored--it's removal won't change that.

> As an alternative, will it make sense to override the RuntimeFeatureFlag? this way one can have multiple paths enabled/disabled by setting a single flag.

That's exactly what this PR does. `CompressorPerRoute.response_compression_enabled` overrides both the (deprecated) `runtime_enabled` flag and the `CommonDirectionConfig.enabled` flag. It makes sense for the config with the narrower scope  to override the config with the broader listener-wide scope.

Reply:
> That's exactly what this PR does. `CompressorPerRoute.response_compression_enabled` overrides both the (deprecated) `runtime_enabled` flag and the `CommonDirectionConfig.enabled` flag. It makes sense for the config with the narrower scope to override the config with the broader listener-wide scope.

Sorry, I wasn't clear before. What I meant is that there will be a per-route override of [`enabled`](https://github.com/envoyproxy/envoy/blob/e4ae2259add5a327c4d67de18a82df5d78a801b4/api/envoy/extensions/filters/http/compressor/v3/compressor.proto#L34) for the request and the response.
That way one can set a runtime flag name for the request of route R1, and a different runtime flag name for the response of route R2.

Reply:
Are you suggesting something like this?
```
message CompressorPerRoute {
  CommonDirectionConfig request_direction_config = 1;
  CommonDirectionConfig response_direction_config = 2;
}
```

If so, I have two concerns:
1. I don't think anyone actually wants runtime flags for per-route compressor filter enablement.
2. If we use runtime flags, then the config required for each route would be much bigger. For deployments with lots of routes, this could be a major issue.

Reply:
For concern 1, I assume that the fields were introduced in #14302 for a valid use-case, but not sure if they are currently used.
I fully agree with concern 2, and for this reason I think a `disabled` flag (similar to [ext_authz](https://github.com/envoyproxy/envoy/blob/e4ae2259add5a327c4d67de18a82df5d78a801b4/api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto#L304), [ext_proc](https://github.com/envoyproxy/envoy/blob/e4ae2259add5a327c4d67de18a82df5d78a801b4/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L171)) should be the way to disable this, as it is consistent with other APIs.
My concern is that having an override that flips between the compression between true and false will make the config harder to reason about.

That said, your use-case may be a reasonable where a new approach should be taken.
cc'ing @htuch for more thoughts.

Reply:
The [ext_proc](https://github.com/envoyproxy/envoy/blob/e4ae2259add5a327c4d67de18a82df5d78a801b4/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L171) approach seems reasonable. Applying it here leads to:

```
message CompressorPerRoute {
  oneof override {
    bool disabled = 1 [(validate.rules).bool = {const: true}];
    Compressor overrides = 2;
  }}
```

We don't have a requirement to support overriding all of Compressor's config, but I guess we can do it anyway. Does that work for you?

Reply:
Yes, I don't think we need to add an override of the compressor config at this stage.
We can add this later if needed.

Reply:
I copied the oneof + disabled + overrides approach that seems to be the dominant way filters do this. Thank you for explaining what was needed.